service/notification/NotificationReceiver.php

```diff
@@ -0,0 +1,391 @@
+            // the Adyen Customer Area
+            if ($this->isTestNotification($cronCheckTest)) {
+                $unprocessedNotifications = $this->getUnprocessedNotifications();
+                if ($unprocessedNotifications > 0) {
+                    $acceptedMessage .= "\nYou have $unprocessedNotifications unprocessed notifications.";
+                }
+            }
+
+            $this->helperData->adyenLogger()->logDebug('The result is accepted');
+            return $this->returnAccepted($acceptedMessage);
+        } else {
+            $message = 'Mismatch between Live/Test modes of PrestaShop store and the Adyen platform';
+            $this->helperData->adyenLogger()->logError($message);
+            return json_encode(
+                [
+                    'success' => false,
+                    'message' => $message
+                ]
+            );
+        }
+    }
+
+    /**
+     * HTTP Authentication of the notification
+     *
+     * @param $response
+     * @return bool
+     * @throws MerchantAccountCodeException
+     * @throws AuthenticationException
+     * @throws HMACKeyValidationException
+     * @throws \Adyen\AdyenException
+     */
+    protected function authorised($response)
+    {
+        $internalMerchantAccount = $this->merchantAccount;
+        $submittedMerchantAccount = $response['merchantAccountCode'];
+
+        $isTestNotification = $this->isTestNotification($response['pspReference']);
+        if (empty($submittedMerchantAccount) && empty($internalMerchantAccount)) {
+            if ($isTestNotification) {
+                throw new MerchantAccountCodeException('merchantAccountCode is empty in PrestaShop settings');
+            }
+            return false;
+        }
+
+        // validate username and password
+        if ((!isset($_SERVER['PHP_AUTH_USER']) && !isset($_SERVER['PHP_AUTH_PW']))) {
+            if ($isTestNotification) {
+                $message = 'Authentication failed: PHP_AUTH_USER and PHP_AUTH_PW are empty.';
+                $this->helperData->adyenLogger()->logError($message);
+                throw new AuthenticationException($message);
+            }
+            return false;
+        }
+
+        // validate hmac
+
+
+        if (!$this->hmacSignature->isValidNotificationHMAC($this->notificationHMAC, $response)) {
+            $message = 'HMAC key validation failed';
+            $this->helperData->adyenLogger()->logError($message);
+            throw new HMACKeyValidationException($message);
+        }
+
+        $usernameCmp = strcmp($_SERVER['PHP_AUTH_USER'], $this->notificationUsername);
+        $passwordCmp = strcmp($_SERVER['PHP_AUTH_PW'], $this->notificationPassword);
+        if ($usernameCmp === 0 && $passwordCmp === 0) {
+            return true;
+        }
+
+        // If notification is test check if fields are correct if not return error
+        if ($isTestNotification) {
```

Comment:
Can the code reach here?
In the line 186 we check
```
if ($usernameCmp === 0 && $passwordCmp === 0) {
            return true;
        }
```
which means if it is true then we return and never reach this point. Am I missing something here?

Reply:
It can. If the server is misconfigured, we log the error for test notifications.